### DISCUSSION TITLE
Enforcing a unmappedTargetPolicy.

### DISCUSSION THREAD

**Discussion by yogeshdengle** [mapstruct/mapstruct#3210] [Q&A] [answered]
Hello,

In a gradle project I am trying to enforce `unmappedTargetPolicy`.
Basically I want the `unmappedTargetPolicy` to be set to `org.mapstruct.ReportingPolicy#ERROR` and that individual mappers should not be able to override this.
They should just use the `ignore` flag explicitly to mark certain fields as ignored.

I went through the docs and even looked at the compile time args that can be setup; however I did not find what I am looking for.
Also [Shared Configurations](https://mapstruct.org/documentation/stable/reference/html/#shared-configurations) sections talks about the fact that the `Attributes specified in @Mapper take precedence over the attributes specified via the referenced configuration class` 

Bottom line, I believe the framework doesn't support what I need and just to check with the community if I havent missed anything.

TIA!

**Comment by filiphr** [comment, marked as answer]
Indeed this is not possible out-of-the-box with MapStruct.

However, what you can do to enforce this in your own projects is to implement your own annotation processor that is going to process the `@BeanMapping`, `@Mapper` and `@MapperConfig` annotations and produce a compiler error if someone has defined the `unmappedTargetPolicy`. An annotation processor can determine if a value in an annotation is set by the user or its value is the default value.

**Comment by jvwilge** [comment]
Another option is to use ArchUnit (or a similar tool). If you're getting this strict there are probably many other things you want to enforce with ArchUnit too. In ArchUnit it would look like this :

```java
    @ArchTest
    ArchRule MAPPER_RULE = ArchRuleDefinition.classes().that().areAnnotatedWith(Mapper.class)
            .should(notHaveDefinedUnmappedTargetPolicy());

    private ArchCondition<? super JavaClass> notHaveDefinedUnmappedTargetPolicy() {
        return new ArchCondition<>("not define unmappedTargetPolicy") {

            @Override
            public void check(JavaClass item, ConditionEvents events) {
                final JavaAnnotation<JavaClass> javaAnnotation = item.getAnnotationOfType("org.mapstruct.Mapper");

                long errors = javaAnnotation.get("unmappedTargetPolicy")
                        .map(Object::toString)
                        .filter(r -> !r.equals("ReportingPolicy.IGNORE"))
                        .stream().count();

                if (errors > 0) {
                    events.add(new SimpleConditionEvent(item.getSimpleName(), false, "the unmappedTargetPolicy should be removed"));
                } else {
                    events.add(new SimpleConditionEvent(item.getSimpleName(), true, "ok"));
                }
            }
        };
    }
```

Technically you can still override it with IGNORE, but it gets the desired effect.

**Reply by filiphr** [reply]
The solution I suggested with a custom annotation processor is similar to this. You only get a compile error versus running a test.

**Comment by yogeshdengle** [comment]
Sorry folks for some reason this was picked up by an email filter and I didnt get notifications for this. Nevertheless both are great options. Will see what works for us. Thank you!
